spring-integration-core/src/main/java/org/springframework/integration/transformer/ToProtobufTransformer.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright 2023-2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.integration.transformer;
+
+import org.springframework.integration.transformer.support.ProtoHeaders;
+import org.springframework.messaging.Message;
+import org.springframework.messaging.MessageHeaders;
+import org.springframework.messaging.converter.ProtobufMessageConverter;
+import org.springframework.messaging.support.MessageHeaderAccessor;
+import org.springframework.util.Assert;
+
+/**
+ * An Protocol Buffer transformer for generated {@link com.google.protobuf.Message} objects.
+ *
+ * @author Christian Tzolov
+ * @since 6.1
+ */
+public class ToProtobufTransformer extends AbstractTransformer {
+
+    private final ProtobufMessageConverter protobufMessageConverter;
+
+    public ToProtobufTransformer() {
+        this(new ProtobufMessageConverter());
+    }
+
+    public ToProtobufTransformer(ProtobufMessageConverter protobufMessageConverter) {
+        this.protobufMessageConverter = protobufMessageConverter;
+    }
+
+    @Override
+    protected Object doTransform(Message<?> message) {
+        Assert.isInstanceOf(com.google.protobuf.Message.class, message.getPayload(),
+                "Payload must be an implementation of 'com.google.protobuf.Message'");
+
+        MessageHeaderAccessor accessor = new MessageHeaderAccessor(message);
+        accessor.setHeader(ProtoHeaders.TYPE, message.getPayload().getClass().getName());
+        if (!message.getHeaders().containsKey(MessageHeaders.CONTENT_TYPE)) {
+            accessor.setContentType(ProtobufMessageConverter.PROTOBUF);
+        }
+
+        return this.protobufMessageConverter.toMessage(message.getPayload(), accessor.getMessageHeaders());
```

Comment:
How about that `textToBytes` option we have chatted before?

Reply:
I thought about it, but, IMO, the right solution is to override the ProtobufMessageConverter#convertToInternal() method and make it return Message<byte[]> instead of Message<String>.
It would be a waste to get back a Message<String> (from the ProtobufMessageConverter) just rewrite it on the fly into  Message<byte[]>.
What do you think?

Reply:
I've tried to override the default [ProtobufMessageConverter](https://github.com/spring-projects/spring-framework/blob/main/spring-messaging/src/main/java/org/springframework/messaging/converter/ProtobufMessageConverter.java) JSON handling. 
In `ToProtobufTransformer` I added:
```java
    public static class ProtobufMessageConverterEx extends ProtobufMessageConverter {
        @Override
        protected Object convertToInternal(Object payload, MessageHeaders headers, Object conversionHint) {
            Object outputPayload = super.convertToInternal(payload, headers, conversionHint);
            return (outputPayload instanceof String) ? ((String) outputPayload).getBytes() : outputPayload;
        }
    }
```
It does the string-to-byte conversion but then if you try to transform it back (using the `FromProtobufTransformer` and in turn the `ProtobufMessageConverter`) I we get stuck with the 
[ProtobufMessageConverter.ProtobufJavaUtilSupport#merge(...)](https://github.com/spring-projects/spring-framework/blob/88e6544d9d3498359b49a2548a1b6ef52b32a5ad/spring-messaging/src/main/java/org/springframework/messaging/converter/ProtobufMessageConverter.java#L268) that assumes `String` payload only and fails if the payload is `byte[]`.
IMO there is no way to override this behaviour. If we want to pursue this path we need to fix the ProtobufMessageConverter impl. to use the `merge(Reader...` instead of `merge(String ...`

So to me it seems that we should stick with the String json representation, but document it well and explain how to perform `textToBytes` & `bytesToText` transformations as additional processing steps.

Reply:
Yes, fully agree: since there is no way for non-String JSON handling in the `ProtobufMessageConverter`, then for consistency our transformers will stick with the same behavior.